extensions/rust/src/route_map.rs

```diff
@@ -0,0 +1,454 @@
+use crate::util::{get_base_components, path_parameters_eq};
+
+use std::collections::{hash_map, HashMap, HashSet};
+
+use pyo3::{
+    prelude::*,
+    types::{PyDict, PyFunction, PyTuple, PyType},
+};
+
+pyo3::import_exception!(starlite.exceptions, ImproperlyConfiguredException);
+pyo3::import_exception!(starlite.exceptions, MethodNotAllowedException);
+pyo3::import_exception!(starlite.exceptions, NotFoundException);
+
+/// A context object that stores Python handles that are needed in the trie
+struct StarliteContext {
+    /// HTTPRoute
+    http_route: Py<PyType>,
+    /// WebSocketRoute
+    web_socket_route: Py<PyType>,
+    /// ASGIRoute
+    asgi_route: Py<PyType>,
+    /// starlite.parsers.parse_path_params
+    parse_path_params: Py<PyFunction>,
+}
+
+/// A node for the trie
+#[derive(Debug, Clone)]
+struct Node {
+    components: HashSet<String>,
+    children: HashMap<String, Node>,
+    path_parameters: Option<Vec<HashMap<String, Py<PyAny>>>>,
+    asgi_handlers: Option<HashMap<String, Py<PyAny>>>,
+    is_asgi: bool,
+    static_path: Option<String>,
+}
+
+impl Node {
+    /// Creates a new trie node
+    pub fn new() -> Self {
+        Self {
+            components: HashSet::new(),
+            children: HashMap::new(),
+            path_parameters: None,
+            asgi_handlers: None,
+            is_asgi: false,
+            static_path: None,
+        }
+    }
+
+    /// Converts the Rust representation to a PyDict representation
+    pub fn as_pydict(&self) -> PyResult<Py<PyDict>> {
+        let gil = Python::acquire_gil();
+        let dict = PyDict::new(gil.python());
+
+        dict.set_item("components", self.components.clone())?;
+        if let Some(ref asgi_handlers) = self.asgi_handlers {
+            dict.set_item("asgi_handlers", asgi_handlers)?;
+        }
+
+        if let Some(ref path_parameters) = self.path_parameters {
+            dict.set_item("path_parameters", path_parameters)?;
+        }
+
+        dict.set_item("is_asgi", self.is_asgi)?;
+
+        if let Some(ref static_path) = self.static_path {
+            dict.set_item("static_path", static_path)?;
+        }
+
+        Ok(dict.into())
+    }
+}
+
+/// A path router based on a prefix tree / trie.
+///
+/// Stores a handler references and other metadata for each node,
+/// as well as both static paths and plain routes for use with Starlite.
+///
+/// Routes can be added using `add_routes`.
+/// Given a scope containing a path, can retrieve handlers using `parse_scope_to_route`.
+#[pyclass]
+pub struct RouteMap {
+    map: Node,
+    static_paths: HashSet<String>,
+    plain_routes: HashSet<String>,
+    starlite: Py<PyAny>,
+    ctx: StarliteContext,
+}
+
+// The functions below are available to Python code
+#[pymethods]
+impl RouteMap {
+    /// Creates an empty `RouteMap`
+    #[new]
+    pub fn new(starlite: Py<PyAny>) -> PyResult<Self> {
```

Comment:
The first non-self argument to `pyo3` methods can be a `Python<'_>`, so you don't have to go through `Python::acquire_gil().python()`

Reply:
Fixed, thanks